Feature/history

SeokHoChoi opened the pull request:
- routes 파일 임시 생성 (현재까진 역할이 없습니다)
- src/data 폴더에 json 형식의 mockData 생성 (현재 mockDataDetail1.json 으로 테스트 중입니다)
- tech-detail-page 생성 
- 초기 상태와 통신 후 리렌더링을 위한 _setupInitialState 메서드 생성

현재까진 디테일 페이지를 위한 id 값을 _setupInitialState에서 const id = 1; 을 이용해 임시로 통신 테스트 진행하였습니다!

SeokHoChoi committed:
feat: routes 와 tech-detail-page 생성

SeokHoChoi committed:
feat: tech mockData 추가

SeokHoChoi committed:
feat: detail page - router를 통한 이동 없는 임시 완성

SeokHoChoi committed:
feat: tech-blog-page.js 파일에 fetch 통신 코드와 HTML 템플릿 추가

f-lab-jeremy commented on src/pages/tech-blog-page.js:
```diff
@@ -1,19 +1,38 @@
-        </div>
-      `;
+  _setupInitialState() {
+    fetch('src/data/mockDataOfTech.json')
+      .then((response) => response.json())
+      .then((data) => {
+        this._$state = data;
+        // 상태 변경을 반영하기 위해 실행
+        this._render();
+      });
```
이런 fetch 로직은 바로 가져다 사용하기보단 한번 추상화해서 사용하면 좋을 것 같아요.
나중에 네트워크 통신 로직을 fetch가 아니라 XMLHttpRequest로 바꾼다면? axios로 바꾼다면?

SeokHoChoi replied:
api 폴더에 getMockData를 만들어 tech-blog-page 자체에서는 fetch에서 axios로 변경하여도 영향이 가지 않도록 하였습니다!

f-lab-jeremy replied:
[cf7979e](https://github.com/SeokHoChoi/pure-js-spa/pull/6/commits/cf7979e5c03edb17db2cbc2517af924b5428f93d) 이렇게되면 API가 추가될때마다 저런 함수가 추가될거고, 그럼 결국 axios로 바꾼다면 모든 곳을 변경해야하는건 동일해지지 않을까요?
아예 네트워크 통신 부분을 하나의 인터페이스로 추상화해볼 수 있을 것 같아요. 아마 책에도 나와있을겁니다!

SeokHoChoi replied:
말씀 감사합니다!! 커밋 메시지에 상세히 작성하여 다시 푸시 하였습니다!

f-lab-jeremy commented on src/pages/tech-detail-page.js:
```diff
@@ -0,0 +1,30 @@
+      .then((data) => {
+        this._$state = data;
+        // 상태 변경을 반영하기 위해 실행
+        this._render();
```
지금 보면 Component라는 슈퍼 클래스에 _render 메서드가 존재하고, 이 _render 메서드는 서브클래스의 _template 메서드를 호출하여 렌더링을 하는 것 같은데, 이렇게 서브 -> 슈퍼 -> 서브로 호출되도록 구조를 짜신 이유가 있으실까요?
Component 클래스의 책임은 뭐라고 생각하면서 코드를 작성하셨는지 설명해주세요 ㅎㅎ

SeokHoChoi replied:
현재 Component 클래스는 서브 클래스의 뼈대(인터페이스) 역할을 해주도록 만들었습니다!
_render 메서드가 _template 메서드를 호출하는 상황에서
지금처럼 서브 클래스에서 다시 _render 메서드가 호출되는 상황을 _$state의 변경을 감지하는 메서드를 만들어
서브 클래스에서의 _render 메서드 호출 없이 리렌더링을 하려 합니다!

f-lab-jeremy reviewed (commented):
- 제가 달아드리는 코드리뷰는 정답이라고 단정짓지 마시고, 비판적으로 석호님이 진짜 이게 맞나? 고민하면서 수용해주세요!
- 커밋 메세지에는 이런 작업을 한 "이유"에 대해 적어주시면 더 좋을 것 같아요. 무슨 작업을 했는지는 사실 코드를 봐도 알거라서 이유가 더 중요합니다! 왜 했는지
- history 브랜치인데 정작 라우팅 관련 코드는 없네요... ㅎㅎ

f-lab-jeremy commented on src/pages/tech-blog-page.js:
```diff
@@ -1,19 +1,38 @@
+    fetch('src/data/mockDataOfTech.json')
+      .then((response) => response.json())
+      .then((data) => {
+        this._$state = data;
+        // 상태 변경을 반영하기 위해 실행
+        this._render();
```
어차피 항상 state가 변경되면 render가 호출돼야할 것 같은데, 이걸 어떻게 자동화 할 수 있을까요? 저희 책에 나온 옵저버 패턴을 적용해보셔도 좋을 것 같아요

SeokHoChoi committed:
feat: 네트워크 통신 로직 변경대비 추상화

SeokHoChoi committed:
feat: 네트워크 통신 추상화

네트워크 통신 부분을 추상화하여 일반적으로 사용할 수 있는 인터페이스를 도입했습니다. 이로써 미래에 API가 추가될 때마다 변경을 최소화하고, 나중에 axios와 같은 다른 라이브러리로 전환해도 전체 코드를 수정할 필요가 없습니다.

-  함수를 도입하여 HTTP 요청을 보내는 일반적인 메서드로 추상화
-  함수를 통해 간편하게 GET 요청을 보낼 수 있도록 지원

이 변경으로 코드 유지보수성이 향상되었습니다.

SeokHoChoi committed:
feat: 프래그먼트 식별자 없이 경로 정의를 위한 라우터 생성

SeokHoChoi committed:
feat:상태 변경 시 리렌더링 자동화를 위한 _setState 메서드와 이벤트 설계와 등록을 위한 _setEvent 메서드와 _addEvent 메서드 추가

SeokHoChoi committed:
feat: 페이지 전환을 위한 라우터 기능 추가

SeokHoChoi committed:
refactor: 기본경로 변경

f-lab-jeremy commented on src/api/mockApi.js:
```diff
@@ -0,0 +1,40 @@
+  const { status } = response;
+  let data;
+  // 클라이언트의 요청은 정상적이나 컨텐츠를 제공하지 않는경우(HTTP Response body가 존재하지 않습니다)
+  if (status !== 204) {
+    data = await response.json();
+  }
```
이러면 400, 500 에러일때도 문제 없이 동작해서 의도치않은 결과가 생길 것 같네요!

SeokHoChoi replied:
감사합니다 feature/typescript 브랜치에서 수정 했습니다!

f-lab-jeremy commented on src/api/mockApi.js:
```diff
@@ -0,0 +1,40 @@
+    headers: new window.Headers(headers),
+  };
+
+  if (body) {
+    config.body = JSON.stringify(body);
+  }
```
json body를 넣으려면 헤더에 추가적으로 설정해줘야햐 하는 속성이 있어요. 이렇게만하면 서버에서 못읽을거에요!

SeokHoChoi replied:
감사합니다!
혹시 config.headers['Content-Type'] = 'application/json'; 처리가 맞을까요??
현재 request에서 headers를 파라미터로 받고 있어 config.headers['Content-Type'] = 'application/json'; 가 별도로 필요 없다고 판단했습니다!

f-lab-jeremy commented on src/core/component.js:
```diff
@@ -45,6 +45,8 @@ export default class Component {
    * _componentDidUpdate 메서드 에서의 this._$target = this._template() 실행 코드의 중복을 제거해 줍니다.
    */
   _render() {
+    console.log(this._$target, '음..');
```
이런 로그코드는 빼고 올려주세요!

SeokHoChoi replied:
넵 감사합니다!!

f-lab-jeremy commented on src/routes/router.js:
```diff
@@ -0,0 +1,88 @@
+  let notFound = () => {};
+  let lastPathname;
+
+  const router = {};
```
이거 직접 구현하신걸까요 아니면 책을 참고해서 구현하신걸까요? 저번에 말씀드렸듯이 책을 참고할꺼면 어떻게 구현할까 고민을 많이 해보고 참고하시면 더 좋을 것 같아요~!
일단 무작정 따라치기보단 고민 이후에 참고하는게 더 와닿으실거에요

SeokHoChoi replied:
- 멘토링때 상담 완료

f-lab-jeremy commented on src/core/component.js:
```diff
@@ -55,4 +55,31 @@ export default class Component {
+   */
+  _addEvent(eventType, selector, callback) {
+    this._$target.addEventListener(eventType, (event) => {
+      if (!event.target.closest(selector)) return false;
```
`if (!event.target.closest(selector)) return false`
이 로직은 왜 넣으셨나요?

SeokHoChoi replied:
가장 가까운 상위 요소를 선택하기 위해 넣었습니다!

브라우저 라우터로 변경한 이번 PR 에서도 이벤트 위임을 위해 활용하였습니다!

f-lab-jeremy commented on src/core/component.js:
```diff
@@ -55,4 +55,31 @@ export default class Component {
+   * @param {*} newState - 기존의 상태를 새로운 상태로 변경할 상태가 들어옵니다.
+   * @description 상태를 변경 후 _render() 메서드를 호출하여 브라우저 출력 내용을 state에 종속시킵니다.
+   */
+  _setState(newState) {
+    this._$state = { ...this._$state, ...newState };
+    this._render();
+  }
```
👍

f-lab-jeremy commented on src/pages/index.js:
```diff
@@ -0,0 +1,14 @@
+import TechBlogPage from './tech-blog-page.js';
+import TechDetailPage from './tech-detail-page.js';
```
import할 때 js 확장자를 안붙이면 어떻게 될까요? 왜 꼭 붙여야할까요? 분명 CRA로 react 프로젝트 파서 할 때는 안붙여도 잘 됐었는데요

SeokHoChoi replied:
동작하지 않습니다!
현재 SPA 구현을 위해 type="module"을 사용하고 있습니다.

ES6 모듈 시스템에서 모듈을 가져올때 파일 확장자를 명시해야 합니다.
명시 되어야 모듈이 정확히 어떤 파일을 가져올지 런타임에 알려주는 역할을 합니다!

CRA로 리액트 프로젝트를 만들시 Webpack을 내장하고 있어 resolve를 이용하여 확장자를 명시하지 않아도 파일을 찾아내고 처리할 수 있습니다!

SeokHoChoi committed:
feat:  브라우저 라우터로 변경하여 SPA 프로젝트 개선

해시 라우터의 서버 요청의 한계와 URL 가독성이 떨어지는 문제가 있었습니다. 브라우저 라우터를 도입함으로써 검색 엔진 최적화가 가능해지고, 깔끔한 URL을 유지할 수 있어 사용자 경험을 향상시켰습니다.

SeokHoChoi committed:
refactor: 스크립트 또는 모듈이 실행되는 위치에 따라 경로가 동적으로 변할 수 있기에 get 요청 경로 수정

f-lab-jeremy commented on src/routes/router.js:
```diff
@@ -12,51 +12,84 @@ export default class Router extends Component {
     // 초기 라우팅 체크
     this.checkRoutes();
+
+    document.body.addEventListener('click', (e) => {
+      // closest 메소드를 사용하여 가장 가까운 'a[data-navigation]' 상위 요소를 찾을 수 있습니다.
+      const closestLink = e.target.closest('a[data-navigation]');
+
+      if (closestLink) {
+        e.preventDefault();
+        this.navigate(closestLink.href);
+      }
+    });
```
이 코드는 동작하긴 하겠지만, 라우팅을 트리거 하는곳(data-navigation 속성이 있는 엘리먼트)의 코드와 실제 로직이 호출되는 코드(지금 이 코드)가 물리적으로 멀어서 별로 코드는 아니라고 생각해요.
예를들어 누군가 석호님의 코드를 유지보수하면서 라우팅 로직을 수정하려고 할때, 모든 코드를 파악하지 않는 한 이 router.js 파일의 존재를 알진 못할테니 data-navigation 속성이 붙은 엘리먼트를 먼저 발견할텐데요. 이 속성만 붙어있을뿐 라우팅 관련 코드는 해당 코드 근처에 존재하지 않다보니 라우팅이 동작하는게 마치 매직처럼 느껴질거에요. 그래서 열심히 코드를 뒤지다가 이 router.js를 발견하게 될거구요. 이런 패턴은 별로 좋지 않아요. 보통 "코드의 응집도를 높여야한다"라는 말을 많이 하는데 이런 관점을 고려해서 어떻게 개선할 수 있을지 고민해보셔도 좋을 것 같아요.

저번에 어떤 기능을 내 스스로 한번 구현해보고, 라이브러리는 어떻게 인터페이스를 노출하고 구현했는지 보면 많은 공부가 된다고 말씀드렸었는데요. 같은 맥락으로 react-router가 어떻게 인터페이스를 노출하는지(어떻게 개발자로 하여금 react-router의 기능을 사용할 수 있도록 하는지) 보시면 공부가 되실거에요!
react-router에서는 라우팅할때 이런 방식으로 구현되어있지 않죠. 라우팅을 트리거하는 엘리먼트와 라우팅을 시키는 함수가 근처에 존재하도록 인터페이스를 제공해요.

그리고 이렇게 구현되면 push는 동작하겠지만 pop은 어떻게 구현할 수 있을까요? replace는?

SeokHoChoi replied:
말씀 감사합니다!!! 이해했습니다!! 
"url 변경을 감지하여 컴포넌트를 교체한다"만 계속 생각했었는데 다른 동작들 이나 유지보수까지 생각하지 못했네요..!
react-router 확인하며 다시 고민해 보겠습니다!!

SeokHoChoi committed:
기기 변경에 따른 커밋 테스트
